Oxidize TwoQubitWeylDecomposition

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit is part 1 of migrating the default 2q unitary synthesis to leverage parallel rust described in #8774, the eventual goal is to be able to run unitary synthesis in parallel for all the unitary matrices in a circuit in a single call from the `UnitarySynthesis` pass. This commit lays the initial groundwork for doing this by starting with the largest piece of the default 2q unitary synthesis code, the TwoQubitWeylDecomposition class. It migrates the entire class to be a pyclass in rust. There is still a Python subclass for it that handles the actual QuantumCircuit generation and also the string representations which are dependent on circuit generation. The goal of this is to keep the same basic algorithm in place but re-implement as-is in Rust as a common starting point for eventual improvements to the underlying algorithm as well as parallelizing the synthesis of multiple 2q unitary matrices.

### Details and comments

TODO:

- [x] Fix test failures (this should be a drop in replacement)
  - [x]  Restore specialization tests using rust specialization enum instead of `isinstance` check. These were initially removed because the specialized subclassing mechanism has been removed, but we shouldn't lose the coverage.
- [x] Figure out path to raise `QiskitError` from rust code.
- [x] Benchmark, profile, and optimize after tests are all passing
- [x] Add release note

mtreinish committed:
Oxidize TwoQubitWeylDecomposition

This commit is part 1 of migrating the default 2q unitary synthesis to leverage
parallel rust described in #8774, the eventual goal is to be able to
run unitary synthesis in parallel for all the unitary matrices in a
circuit in a single call from the `UnitarySynthesis` pass. This commit
lays the initial groundwork for doing this by starting with the largest
piece of the default 2q unitary synthesis code, the
TwoQubitWeylDecomposition class. It migrates the entire class
to be a pyclass in rust. There is still a Python subclass for it that
handles the actual QuantumCircuit generation and also the string
representations which are dependent on circuit generation. The goal of
this is to keep the same basic algorithm in place but re-implement
as-is in Rust as a common starting point for eventual improvements to
the underlying algorithm as well as parallelizing the synthesis of
multiple 2q unitary matrices.

ShellyGarion commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +263,792 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+    let m = (a + b + c) / 3.;
+    let [am, bm, cm] = [a - m, b - m, c - m];
+    let [ab, bc, ca] = [a - b, b - c, c - a];
+    m + am * bm * cm * (6. + ab * ab + bc * bc * ca * ca) / 18.
```
as @levbishop pointed out there is a misprint in this formula.
It should be:
```suggestion
    m + am * bm * cm * (6. + ab * ab + bc * bc + ca * ca) / 18.
```
See https://github.com/Qiskit/qiskit/pull/11953

mtreinish committed:
Fix typo in closest_partial_swap()

This commit fixes a typo the formula in the function.
This is the same fix from #11953.

Co-authored-by: Shelly Garion <46566946+ShellyGarion@users.noreply.github.com>

mtreinish committed:
Run black and lint

mtreinish committed:
Fix potential imaginary component sign flip in determinant

mtreinish committed:
Run cargo fmt

mtreinish committed:
Simplify using numpy eigh example comment

mtreinish committed:
Add missing checks to decompose_two_qubit_product_gate()

mtreinish committed:
Use rng first trial from Python implementation

To aid in debugging and rule out rng differences causing different
results this commit switches the first iteration of the randomized loop
to have hard coded values that are identical to what the rng in numpy
was returning. It is very unlikely that this will have any impact
because the specific random numbers used shouldn't matter, this is
mostly to just rule it out as a possibility in debugging the remaining
test failures.

mtreinish committed:
Fix whitespace in error message

mtreinish committed:
Fix assert check

mtreinish committed:
Fix missing multiplier on specialized weyl gate

mtreinish committed:
Fix various mistakes

This commit fixes two fundamental issues in the code. The first is the
rz and ry matrix were being incorrectly constructed for a given angle.
This caused the specializations that were computing the 1q matrices in
the decomposition based on a product with these gates' matrices to
return invalid results. The second issue is for the MirrorControlledEquiv
specialization had the angles backwards for computing the matrix of the
rz gates used in the products for the 1q matrices:

`K1l = K1l @ Rz(K1r)` and `K1r = K1r @ Rz(K1l)` not
`K1l = K1l @ Rz(K1l)` and `K1r = K1r @ Rz(K1r)`

This was a typo from the original transcription.

mtreinish committed:
Add pickle serialization support for TwoQubitWeylDecomposition

mtreinish committed:
Remove duplicate conjs increment

mtreinish committed:
Fix fSim abmb equivalent specialization

mtreinish committed:
Add specialized test class back

mtreinish committed:
Use QiskitError for backwards compatibility

mtreinish committed:
Add release note

mtreinish committed:
Make explicit specialization private

mtreinish commented:
There's still potential work to do on benchmarking, profiling, and tuning the implementation here. But I've marked it ready for review now because the majority of this PR is finished. As there weren't expected to be huge performance wins from just this PR (it was more about enabling future performance improvements by using rust for the dependents of this class) the potential tuning isn't as critical now and we can always work on it in a follow up if needed.

jakelishman commented on crates/accelerate/src/euler_one_qubit_decomposer.rs:
```diff
@@ -553,7 +553,7 @@ pub fn generate_circuit(
 #[inline]
-fn angles_from_unitary(unitary: ArrayView2<Complex64>, target_basis: &str) -> [f64; 4] {
+pub fn angles_from_unitary(unitary: ArrayView2<Complex64>, target_basis: &str) -> [f64; 4] {
     match target_basis {
```
This function contains an `unreachable` which implies conformance with `VALID_BASES` which is a `const` scope to `unitary_to_gate_sequence`.  Since we're making this more public, can we potentially look to change `VALID_BASES` into a full enum, or at least make this method return a `Result` so we can raise up a Python-space error rather than panicking on bad input?

Personally coercing the valid bases to an `enum` on entry from Python space feels cleanest to me (if nothing else, it's way easier and freer to `Copy` no-data enum variants all over the place than to have to work with `str`/`String`), but I can also live with keeping things stringly-typed if we stay diligent about using exceptions.

mtreinish replied:
I migrated to this in: https://github.com/Qiskit/qiskit/pull/11946/commits/33d073d2e05ecaa326a9bf70f02c1d104d29d703 it does kind of feel like a bit of feature creep in this PR though, but I agree it's a cleaner interface for this

jakelishman commented on crates/accelerate/src/euler_one_qubit_decomposer.rs:
```diff
@@ -609,6 +609,7 @@ fn compute_error(
 
+#[inline]
 #[pyfunction]
 pub fn compute_error_one_qubit_sequence(
```
This is purely a `pyfunction` and not called from within Rust - was there a reason to annotate it `inline`?

(Comment repeats in a few other places.)

mtreinish replied:
There isn't I was trying to ensure the public functions we call inside rust were annotated as inline. But you're right this one and most of the others I added `[inline]` to don't do anything and are more confusing than anything else.

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/11946/commits/2c5d9634d5a643f37539a821115ed727504d1a8f

jakelishman commented on crates/accelerate/src/euler_one_qubit_decomposer.rs:
```diff
@@ -27,7 +27,7 @@ use numpy::PyReadonlyArray2;
 use crate::utils::SliceOrInt;
 
-const DEFAULT_ATOL: f64 = 1e-12;
+pub const DEFAULT_ATOL: f64 = 1e-12;
```
This is getting exported, but does it make full sense to link this magic number with the 2q decomposer?  This one is currently used in 1q decomposition sequences, but this PR extends it to be used in 2q sequences as well.

It may well make sense, but if so, perhaps a more descriptive name of what this number is meant to be for would help - `ATOL` is so vague it's really unclear how it should be / will be used, whereas something like `ANGLE_ZERO_EPSILON` would cover its current uses and hopefully discourage us from linking the same magic number to even more unrelated use-cases in the future.

mtreinish replied:
Renamed in: https://github.com/Qiskit/qiskit/pull/11946/commits/376b2e7d9aa3746b6e028697b4b3e84a62f8bf08

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -18,59 +18,119 @@
+    ],
+    [
+        Complex64::new(1., 0.),
+        Complex64::new(-0., -1.),
```
lol at the signed zero here

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/11946/commits/89070e445e4cf60c3e825634ddc139d9c97b5a8d

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -18,59 +18,119 @@
+        Complex64::new(-0.5, -0.),
+        Complex64::new(0., 0.),
+    ],
 ];
```
Ok, this one _really_ confuses me with the presence / absence of the signed zeros.

mtreinish replied:
TBH, I just copied the output from printing the numpy version. I can remove all the negative 0s, I actually meant to do this earlier. This was from when I was bogged down getting distracted by every little small fp difference and was trying to minimize the differences.

mtreinish replied:
I just dropped these in: https://github.com/Qiskit/qiskit/pull/11946/commits/89070e445e4cf60c3e825634ddc139d9c97b5a8d

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -18,59 +18,119 @@
+    .view()
+    .into_faer_complex()
+    .to_owned()
 }
```
Would it make sense to define this in terms of `transform_from_magic_basis`?

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/11946/commits/9bf2f155888399d1686ef0b2cf94c9e229f01720 Although I guess maybe it's better for the faer version to use faer for the matmul, but it doesn't matter a ton

jakelishman replied:
Oh yeah, I was just commenting on the current form, which uses `ndarray` internally, so is entirely the same as `transform_from_magic_basis`.  I'd also have been fine with making this Faer-native.

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+            Specializations::SimaabEquiv => 7,
+            Specializations::SimabbEquiv => 8,
+            Specializations::SimabmbEquiv => 9,
+        };
```
Feels like this serialisation/deserialisation of the `pyclass` enum could be handled by a `pymethod` on the enum itself?

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/11946/commits/5fac9acd72b937b19e5f08fce5305a83682a6890 I didn't use a pythmethod though because the pickling gets a bit odd for enums because they don't really have a constructor.

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -98,9 +158,41 @@ impl Arg for c64 {
+fn decompose_two_qubit_product_gate(
+    unitary: ArrayView2<Complex64>,
+) -> (Array2<Complex64>, Array2<Complex64>, f64) {
+    let mut r: Array2<Complex64> = unitary.slice(s![..2, ..2]).to_owned();
```
Compared to the Python version, this relaxed the name from `special_unitary` to `unitary` - was that intentional?  If not, might be good to keep it as a hint to ourselves, since `special_unitary` implies that the determinant is exactly one, rather than having magnitude one.

mtreinish replied:
Renamed in: https://github.com/Qiskit/qiskit/pull/11946/commits/cbbdb53f6478aad603c47808177fd1bfbd6a2dcd

I only changed it to save some typing. But you're right it's important context to have here.

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -98,9 +158,41 @@ impl Arg for c64 {
+    assert!(
+        det_r.abs() >= 0.1,
+        "decompose_two_qubit_product_gate: unable to decompose: detR < 0.1"
+    );
```
It would perhaps be better if we could return a proper Python exception from this (and elsewhere).

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/11946/commits/cbbdb53f6478aad603c47808177fd1bfbd6a2dcd

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -98,9 +158,41 @@ impl Arg for c64 {
+        "decompose_two_qubit_product_gate: unable to decompose: detR < 0.1"
+    );
+    r.mapv_inplace(|x| x / det_r.sqrt());
+    let r_t_conj: Array2<Complex64> = r.t().mapv(|x| x.conj()).to_owned();
```
This is neither here nor there, but isn't the result of `mapv` already owned?

mtreinish replied:
```suggestion
    let r_t_conj: Array2<Complex64> = r.t().mapv(|x| x.conj());
```

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -98,9 +158,41 @@ impl Arg for c64 {
+    ];
+    let mut temp = kron(&eye, &r_t_conj);
+    temp = unitary.dot(&temp);
+    let mut l = temp.slice_mut(s![..;2, ..;2]).to_owned();
```
Is there a purpose in taking a mutable slice if we're about to clone it anyway?

mtreinish replied:
Nah, this is just left over from an earlier iteration where I was trying to do everything in place. But I missed we returned `l` in the original iteration of this code so we needed to have a copy.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/11946/commits/cbbdb53f6478aad603c47808177fd1bfbd6a2dcd

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -18,59 +18,119 @@
+    .view()
+    .into_faer_complex()
+    .to_owned()
 }
```
If these are gaining a `reverse` argument, we ought to consider the naming - in Python space this function is called `transform_to_magic_basis` and has a `reverse` argument, whereas here it's called `transform_from_magic_basis` and has a `reverse` argument, except the Rust space ones appear to have the same behaviour as the Python-space ones.  For consistency, should probably pick one or the other, or if we're being really extra:
```rust
#[derive(Clone, Copy)]
enum MagicBasisTransform {
   Into,
   OutOf,
}

fn magic_basis_transform(mat: ArrayView2<Complex64>, direction: MagicBasisTransform) {}
```

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/11946/commits/9bf2f155888399d1686ef0b2cf94c9e229f01720

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
 
+fn trace_to_fid(trace: Complex64) -> f64 {
+    (4.0 + trace.abs().powi(2)) / 20.0
+}
```
Do you think it could be any clearer to make `trace_to_fid` a trait that we impl for `Complex64` and `c64`?

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/11946/commits/9a914c2a08fba7597490222aea06698be0b8b299

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+    SimaabEquiv,
+    SimabbEquiv,
+    SimabmbEquiv,
+}
```
On an incredibly minor point of naming: wouldn't a non-flags enum name be singular, not plural?  You can by definition only be referencing a singular one at a time.

mtreinish replied:
Yeah that's fair it probably should be `Specialization` not `Specializations`. I think when I first made the enum I was thinking "These are the specializations".

mtreinish replied:
Renamed in: https://github.com/Qiskit/qiskit/pull/11946/commits/99db4bf986e6ab528ccd27562a6741a26b8a1dca

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+    #[pyo3(get)]
+    calculated_fidelity: f64,
+    unitary_matrix: Array2<Complex64>,
+}
```
I'm a big fan of unifying the specialisations into a single struct like this - it's much more readable to me than all the subclasses, and makes it more obvious how things should be extended and when they're being generalised.

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+    }
+}
+
+const IPZ: [[Complex64; 2]; 2] = [
+    [C1_IM, Complex64::new(0., 0.)],
+    [Complex64::new(0., 0.), Complex64::new(0., -1.)],
+];
```
_Technically_, the Python-space version of this code would have had `Complex64::new(-0.0, -1.0)` in the last field, with a negative zero in the real component.

Honestly, it might be worth avoiding the constants here, just for consistency with the other values.

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+                unitary_matrix: Array2::zeros((0, 0)),
+            });
+        }
+        let ipz: ArrayView2<Complex64> = aview2(&IPZ);
```
Since we've done away with the class structure, might it be easier here to remove the `_pickle_context` stuff and use an alternative constructor in conjunction with `__reduce__` so we don't need to attempt to produce a "zeroed" value?

mtreinish replied:
I wasn't sure of the incantation with `__reduce__` to produce that. I always get confused using `__reduce__` because there are so many different modes of operation for it.  If you've got a pattern that you think works here I'd gladly remove this because this was a really ugly way to avoid doing computation on new object construction.

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+            // let p_inner = p_raw
+            //     .extract::<PyReadonlyArray2<f64>>()?
+            //     .as_array()
+            //     .mapv(Complex64::from);
```
Do you want to leave this in now things are seemingly working?

mtreinish replied:
Oh, I thought I had actually deleted this already. Yeah it's not needed anymore since everything is working with faer now. I actually think some of my comments above about using faer for matmul and the determinant are probably not true anymore. For getting to this point I was getting distracted by all the divergence between the rust and python versions even if they were equivalent because it made it very difficult to track where the mistakes were. It might be worth investigating if there is a performance benefit to using faer for matmul (their benchmarks claim there are: https://faer-rs.github.io/bench.html#matrix-multiplication-c64 but we're also not using the openblas backend, I don't know how the generic fallback matmul in ndarray compares).

mtreinish replied:
```suggestion
```

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+        (order[0], order[1], order[2]) = (order[1], order[2], order[0]);
+        (cs[0], cs[1], cs[2]) = (cs[order[0]], cs[order[1]], cs[order[2]]);
+        (d[0], d[1], d[2]) = (d[order[0]], d[order[1]], d[order[2]]);
+        let mut p_orig = p.clone();
+        for (i, item) in order.iter().enumerate().take(3) {
+            let slice_a = p.slice_mut(s![.., i]);
+            let slice_b = p_orig.slice_mut(s![.., *item]);
+            Zip::from(slice_a).and(slice_b).for_each(::std::mem::swap);
+        }
```
Minor: is there no way in `ndarray` to assign to subslices, like Numpy does? This is rather harder to read than the equivalent Numpy code.

mtreinish replied:
At the time I couldn't find a pattern for doing this and some issues I dug up where people were asking for this suggested a similar pattern to this. But that might have been an old issue because it looks like I can probably do something like: `p.slice_mut(s![.., i]).assign(p.slice(s![..., *item])` (which internally is just doing the same clone and zip basically)

I seem to remember having trouble using `order` in the slice `s!` macro syntax though, so I might keep the for loop. Although that would result in three allocations for the internal clone in `assign()`  instead of just the one outer one.

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+        global_phase += phase_l + phase_r;
+
+        // Flip into Weyl chamber
+        if cs[0] > PI2 {
+            cs[0] -= PI32;
```
I really wouldn't mind getting rid of some of these constants - `PI32` could reasonably be either $3\pi/2$ or $\pi/32$ locally, and since we're now in a compiled language, LLVM should fold constants where appropriate.

That said, I suppose `PI32` has the advantage of ensuring that we're always using the same order of operations.

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+                // Set to false here to default to general specialization in the absence of a
+                // fidelity and provided specialization.
+                None => false,
+            }
```
I found this inner `fidelity` check to be very hard to follow compared to the Python-space logic of specifically short-circuiting the checks if `fidelity` was `None`.  Could we refactor the match on `_specialization` further down to be something like:
```rust
let specialization = match (_specialization, fidelity) {
    (Some(specialization), _) => specialization,
    (None, None) => Specialization::General,
    (None, Some(fidelity)) => {
        let is_close = |ap, bp, cp| -> bool {
            // Same as before, but now closes over the known `f64` of the fidelity.
        };
        if is_close(0.0, 0.0, 0.0) {
            Specialization::IdEquiv,
        } // ...
    }
};
```
(or a couple of if-lets, or an `Option::or` followed by an `Option::or_else`)?

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+            Specializations::PartialSWAPEquiv => {
+                let closest = closest_partial_swap(a, b, c);
+                let mut k2r_temp = K2l.t().to_owned();
+                k2r_temp.view_mut().mapv_inplace(|x| x.conj());
```
This is confusingly named - I guess it's `k2r_temp` because it's used in the `K2r` calcs, but really it's the dagger of `K2l`.  Could we call it `K2l_ct` (or `_dag` or `_d` or whatever you like / think is most consistent)?

mtreinish replied:
Updated in: https://github.com/Qiskit/qiskit/pull/11946/commits/b273c9398a05d3e9c760354adfc0eb80799953e1

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+            Specializations::PartialSWAPFlipEquiv => {
+                let closest = closest_partial_swap(a, b, -c);
+                let mut k2_temp = K2l.t().to_owned();
+                k2_temp.mapv_inplace(|x| x.conj());
```
Same comment about some dagger/conjugate-transpose suffix.

mtreinish replied:
Updated in: https://github.com/Qiskit/qiskit/pull/11946/commits/b273c9398a05d3e9c760354adfc0eb80799953e1

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+                let [k2ltheta, k2lphi, k2llambda, k2lphase] =
+                    angles_from_unitary(K2l.view(), default_euler_basis);
+                let [k2rtheta, k2rphi, k2rlambda, k2rphase] =
+                    angles_from_unitary(K2r.view(), default_euler_basis);
```
Minor: this feels like it could benefit from the allowed Euler basis being an explicit `enum`.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/11946/commits/33d073d2e05ecaa326a9bf70f02c1d104d29d703

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+                requested_fidelity: fidelity,
+                calculated_fidelity: -1.0,
+                unitary_matrix,
+            },
```
With all of these, it's not so immediately obvious to me when things are getting inherited from the defaults, so there's quite a lot of noise.  I've made mtreinish/qiskit-core#23 that's an attempt I'd made to make the inheritance more explicit, but I'm not precious about whether you accept that or not.

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+        Ok(specialized)
+    }
+
+    #[allow(non_snake_case)]
+    #[getter]
+    fn K1l(&self, py: Python) -> PyObject {
+        self.K1l.to_pyarray(py).into()
+    }
+
+    #[allow(non_snake_case)]
+    #[getter]
+    fn K1r(&self, py: Python) -> PyObject {
+        self.K1r.to_pyarray(py).into()
+    }
+
+    #[allow(non_snake_case)]
+    #[getter]
+    fn K2l(&self, py: Python) -> PyObject {
+        self.K2l.to_pyarray(py).into()
+    }
+
+    #[allow(non_snake_case)]
+    #[getter]
+    fn K2r(&self, py: Python) -> PyObject {
+        self.K2r.to_pyarray(py).into()
+    }
+
+    #[getter]
+    fn unitary_matrix(&self, py: Python) -> PyObject {
+        self.unitary_matrix.to_pyarray(py).into()
+    }
```
These feel like potential performance footguns, since they allocate on attribute access.

mtreinish replied:
This should be avoided by: https://github.com/Qiskit/qiskit/pull/11946/commits/a2fcc58a9edd2c1b1253d88a33d8054d0eca47f7 now. We should only be making one copy for the public facing interface now.

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+        atol: Option<f64>,
+    ) -> TwoQubitGateSequence {
+        let binding = self.default_euler_basis.clone();
+        let euler_basis: &str = euler_basis.unwrap_or(&binding);
```
Minor: this is an awkward corner that's wanting an eager clone.  Can we do a delayed initialisation trick like
```suggestion
        let binding;
        let euler_basis = if let Some(basis) = euler_basis {
            basis
        } else {
            binding = self.default_euler_basis.clone();
            &binding
        };
```
to avoid the eager clone, but still extend the lifetime?

Alternatively, is there a need for this to accept `&mut self`?  Nothing looks like it should need to mutate `self` here - `weyl_gate` has the same unnecessary `mut` - and if so, we could skip `binding` and just `let euler_basis = euler_basis.unwrap_or(&self.default_euler_basis);`.

mtreinish replied:
It doesn't need to be `&mut self` anymore, an earlier version of this had this function touching `self.global_phase` instead of the circuit's global phase which was just wrong and the source of a few test failures until I caught it.

mtreinish replied:
Made the borrow immutable and simplified the logic in: https://github.com/Qiskit/qiskit/pull/11946/commits/f2e7b6d669348a2cc86f267270ae6cf8058e13a3

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+        .unwrap();
+        for gate in c2r.gates {
+            gate_sequence.push((gate.0, gate.1, [0, 0]))
+        }
```
This way of doing 1q gates with duplicated indices seems like it could be very bug prone.  We're already using a `Vec` for the parameters - can we just use an explicit `Vec` for the qubits as well?

(notwithstanding potentially switching these known-small allocations to `smallvec` at some point.)

mtreinish replied:
I just switched it up to smallvec and did this in a single go. I only did this to avoid the extra vec overhead because it's only 1 or 2 elements, I forgot about smallvec for some reason. See: https://github.com/Qiskit/qiskit/pull/11946/commits/aff637eb0d830b26d16c978a50e86d52eb06bf2c

jakelishman commented on qiskit/synthesis/two_qubit/two_qubit_decompose.py:
```diff
@@ -96,7 +95,7 @@ def decompose_two_qubit_product_gate(special_unitary_matrix: np.ndarray):
 
 
-class TwoQubitWeylDecomposition:
+class TwoQubitWeylDecomposition(two_qubit_decompose.TwoQubitWeylDecomposition):
```
This might be for a follow-up, but this inheritance-y structure from the inner Rust worker feels very odd to me.  Is there aany plan to change it to either keep the Rust-space decomposer as an internal-only single object with the outer one caching accesses to the fields I noted as potential performance gotchas, or alternatively to do the whole thing in Rust space?

mtreinish replied:
My original thinking for this was I wanted to drop the python side class altogether but it was going to be annoying to handle the `circuits()`, `from_bytes()`, and string/repr methods rust side, so this was what I settled on. I mostly did it this way to avoid duplication on the python side, but I think you're right I'm concerned about the overhead of copying the arrays for the getter on the `.K*` attributes. As cool as I thought it was to get subclassing working like this, I think you're right it'll be better to just wrap the inner rust object here, even if it ends up being a bit more verbose.

mtreinish replied:
I adjusted this in: https://github.com/Qiskit/qiskit/pull/11946/commits/a2fcc58a9edd2c1b1253d88a33d8054d0eca47f7 so we're wrapping the inner rust object now instead of subclassing. That's just a better pattern in this case.

jakelishman commented on releasenotes/notes/rust-two-qubit-weyl-ec551f3f9c812124.yaml:
```diff
@@ -0,0 +1,14 @@
+    the same and were only used for internal dispatch. However, as it was discoverable
+    behavior this release note is to document that this will no longer occur and all
+    instances of :class:`.TwoQubitWeylDecomposition` will be of the same type. There is no
+    change in behavior for public methods of the class.
```
Technically, `TwoQubitWeylDecomposition.specialize` was public before and now doesn't exist.  I mean, it should never have been called by a user, but it technically was public.

(Also, "no longer will self specialize" is a bit of an awkward word order - "will no longer self-specialize" might be a bit easier to parse.)

mtreinish replied:
Oh, that's a good call. We should add it back with the default `NotImplementedError` raise because it no longer does anything. This isn't a change in behavior because we're not subclassing anymore and the default for `TwoQubitWeylDecomposition` was already to raise.

mtreinish replied:
Added the public method back and deprecated it in: https://github.com/Qiskit/qiskit/pull/11946/commits/1a8be8549588fdbeab259e6db5070e016439e086

jakelishman commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +272,919 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+        fidelity: Option<f64>,
+        _specialization: Option<Specializations>,
+        _pickle_context: bool,
+    ) -> PyResult<Self> {
```
Moving this down to Rust-space has meant (evidenced by changes in the test suite) that the public interface to the constructor has changed; now it's only objects that are already interpretable as `PyReadonlyArray2`, whereas before it accepted anything that Numpy could coerce to an array.

I think that's a sufficiently large breaking change that we should take care of it properly - accept `Py<PyAny>` in the constructor, and do the extraction ourselves in Rust space, or use the the Python inheriting wrapper to call `np.asarray(unitary_matrix, dtype=complex)` before it gets passed to here.

mtreinish replied:
That's a good call, we I was thinking of adjusting the python side constructor to do this, but we can easily do this rust side. I guess it really comes down to if we refactor the to just wrap an inner rust private method per: https://github.com/Qiskit/qiskit/pull/11946#discussion_r1520192861 or not. If we do that we should just do this python side.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/11946/commits/a2fcc58a9edd2c1b1253d88a33d8054d0eca47f7

jakelishman reviewed (commented):
Thanks for this!

I very much like the pulling out of the specialisation bits into enum handling - it makes it a lot clearer to me when things are _not_ being overridden and for which specialisations, which I think is a good idea.

I do worry a bit that the main decomposer logic has become even more complicated to read than it already was.  The Numpy-based versions weren't ideal, but they were less verbose and needed fewer temporary variables.  Are you planning anything that might reduce the amount of total complexity from the function here?  Even just some helper functions to better encapsulate the lifetimes of temporaries might do something, because there's loads of new variables with very similar names to previous ones that are hard to keep track of, since they now live for hundreds of lines of code.

I had comments on efficiency, which I've not mentioned inline (unless there are places I forgot).  Just for tracking purposes, here's a top-level summary of them:
- tiny vecs of floats (like for parameter lists) might want to be `SmallVec`, since we never expand beyond a couple of params and we don't really need heap allocations for those
- there are constant matrices kicking around (like identities) that we heap allocate each time, but could make statics
- there seems to be a few places where we call `mapv` on arrays then tack on a `to_owned`: doesn't `mapv` already produce an owned array?
- there's at least one place with chained `map` calls on ndarrays, which is presumably allocating more matrices than we would if we put the logic all in one place.
- in the Python side, we construct some single-gate circuits just to call `Operator` on them, when we could just do `Gate.to_matrix` on the gate objects directly.

jakelishman committed:
Use explicit inheritance from general decomposition (#23)

mtreinish commented:
Just for reference I ran some quick asv benchmarks for benchmarks that I know run this synthesis internally with 6b4501fe (right after I took it out of WIP) and these are the numbers as of that commit:

```
Benchmarks that have improved:                                                                                                                                                                 
                                                                                                                                                                                               
       before           after         ratio                                                                                                                                                    
     [9f228fa4]       [6b4501fe]                                                                                                                                                               
                      <rust-whales-transcription>                                                                                                                                              
-      12.4±0.1ms      11.2±0.04ms     0.90  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(2, 'translator')                                                                
-      17.7±0.2ms       15.7±0.2ms     0.89  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(3, 'translator')                                                                
-      1.63±0.02s          1.43±0s     0.88  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(27, 'translator')                                                               
-        931±10ms          815±6ms     0.88  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(20, 'translator')                                                               
-      52.0±0.4ms       45.1±0.3ms     0.87  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(5, 'translator')                                                                
-         458±2ms        398±0.4ms     0.87  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(14, 'translator')                                                               
-         153±1ms        131±0.7ms     0.86  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(8, 'translator')                                                                
-     8.77±0.04ms      7.44±0.05ms     0.85  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(2, 'synthesis')                                                                 
-      2.20±0.01s          1.87±0s     0.85  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(3)                                                              
-     4.02±0.02ms      3.34±0.01ms     0.83  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(1, 'synthesis')                                                                 
-        882±20ms         731±30ms     0.83  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(14, 'synthesis')                                                                
-         1.80±0s          1.49±0s     0.83  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(2)                                                              
-         220±8ms         181±10ms     0.83  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(8, 'synthesis')                                                                 
-      2.13±0.05s       1.75±0.04s     0.82  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(20, 'synthesis')                                                                
-      4.34±0.04s       3.48±0.02s     0.80  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(27, 'synthesis')                                                                
-         1.40±0s          1.09±0s     0.78  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(0)                                                              
-         840±6ms          532±4ms     0.63  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(1)                                                              
                                                                                                                                                                                               
Benchmarks that have stayed the same:                                                                                                                                                          
                                                                                                                                                                                               
       before           after         ratio                                                                                                                                                    
     [9f228fa4]       [6b4501fe]                                                                                                                                                               
                      <rust-whales-transcription>                                                                                                                                              
transpiler_levels.TranspilerLevelBenchmarks.track_depth_quantum_volume_transpile_50_x_20(3)                                                       
      2.69±0.01ms      2.67±0.02ms     0.99  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(1, 'translator')                                                                
       21.8±0.2ms       21.6±0.2ms     0.99  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 10, 'lookahead')                                                           
          198±2ms        195±0.8ms     0.99  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 100, 'lookahead')                                                          
       23.7±0.2ms      23.4±0.08ms     0.99  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 10, 'decay')                                                               
        209±0.7ms          204±2ms     0.98  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 100, 'decay')                                                              
        49.4±10ms         48.1±4ms     0.97  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(5, 'synthesis')                                                                 
       3.28±0.01s          3.20±0s     0.97  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(1081, 10, 'decay')                                                              
          300±1ms          292±2ms     0.97  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(409, 10, 'decay')                                                               
          317±1ms          302±2ms     0.95  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(409, 10, 'lookahead')                                                           
       2.00±0.02s          1.89±0s     0.94  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(1081, 10, 'lookahead')                                                          
         18.9±3ms         12.1±2ms    ~0.64  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(3, 'synthesis')                                                                 
                                                                                                                                                                                               
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.                                                                                                                                                    
PERFORMANCE INCREASED.
```

mtreinish committed:
Apply suggestions from code review

mtreinish committed:
Use smallvec for circuit sequence params and qubits

jakelishman commented:
Btw, for the efficiency stuff, I wasn't necessarily intending for it to be in this PR (especially given your top comment), I was just noting it so I didn't forget.

mtreinish committed:
Use const 2x2 identity array where possible

mtreinish committed:
circuit() and weyl_gate don't need a mutable self

mtreinish committed:
Remove unnecessary inline annotations

mtreinish committed:
Rename Specializations enum Specialization

mtreinish committed:
Add back specialize() method and deprecate

mtreinish committed:
Reorganize Python/Rust split to wrap rust instead of subclass

This commit reworks the handoff between python and rust space for the
TwoQubitWeylDecomposition class. Previously the Python side
TwoQubitWeylDecomposition was a subclass of the Rust struct/pyclass of
the same name. This was originally done to deduplicate the the getters
for the attributes. However, because of the overhead associated with
some of the rust getters and needing to do some import normalization to
a numpy array the deduplication wasn't worth the cost.

mtreinish committed:
Remove unecessary allocations

mtreinish committed:
Rename DEFAULT_ATOL to ANGLE_ZERO_EPSILON

mtreinish committed:
Stop obsessing over -0

mtreinish committed:
Handle enum to int conversion as method

mtreinish committed:
Cleanup decompose_two_qubit_product_gate()

mtreinish committed:
Use a trait for trace_to_fid()

mtreinish committed:
Use an enum instead of string for euler basis

mtreinish committed:
Fix release note typo

jakelishman commented on releasenotes/notes/rust-two-qubit-weyl-ec551f3f9c812124.yaml:
```diff
@@ -0,0 +1,27 @@
+    :class:`.TwoQubitWeylDecomposition` object. Despite this it was still a
+    documented part of the public API for the class and is now being
+    deprecated without any potential replacement. This release it always will
+    raise a ``NotImplementedError` when called because the specialization
```
```suggestion
    raise a ``NotImplementedError`` when called because the specialization
```

(what the docs build is complaining about)

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/11946/commits/160d70552fcc4a3a29f5d3801849b82b6bebdb20

mtreinish committed:
Improve magic basis transformation functions

Co-authored-by: Jake Lishman <jake.lishman@ibm.com>

mtreinish committed:
Merge branch 'main' into rust-whales-transcription

levbishop commented on qiskit/synthesis/two_qubit/two_qubit_decompose.py:
```diff
@@ -623,22 +272,27 @@ def __init__(self, rxx_equivalent_gate: Type[Gate]):
             except TypeError as _:
                 raise QiskitError("Equivalent gate needs to take exactly 1 angle parameter.") from _
-            decomp = TwoQubitWeylDecomposition(rxx_equivalent_gate(test_angle))
+            decomp = TwoQubitWeylDecomposition(rxx_equivalent_gate(test_angle).to_matrix())
```
Why is this change needed?

jakelishman replied:
Shouldn't be any more, as of a2fcc58a9edd2c1b1253d88a33d8054d0eca47f7: https://github.com/Qiskit/qiskit/pull/11946#discussion_r1520208233

mtreinish replied:
Yeah, this isn't needed anymore, in earlier revisions of the PR we didn't have the `np.asarray()` call wrapping the input to `TwoQubitWeylDecomposition` so things that were arraylike but not actually arrays yet would raise a type error when passed to the rust side. While it doesn't hurt I'll revert this change to minimize the size of the PR to just what's required for this logical change (it's already large enough anyway).

mtreinish replied:
Reverted in: https://github.com/Qiskit/qiskit/pull/11946/commits/27a7587a0c79b6038b07124dca14b7419a43da57

levbishop commented on test/python/synthesis/test_synthesis.py:
```diff
@@ -141,24 +130,10 @@ def check_one_qubit_euler_angles(self, operator, basis="U3", tolerance=1e-14, si
-                logging.DEBUG,
-                msg=f"Unexpected logging entry: {ctx.output[i]}",
-            )
-            self.assertIn("Requested fidelity:", ctx.records[i].getMessage())
```
Removing the debug logging does result in a little less-complete test coverage because now test suite is notchecking that `calculated_fidelity == actual_fidelity` in as many places. Could restore this by adding more `assertAlmostEqual` between `calculated_fidelity` and `actual_fidelity` in the spots where `assertDebugOnly` was previously, or by moving the debug logging into the python `__init__`

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/11946/commits/03c4af53523f4fabbb72d82fe4a0bea61d09ecc4

levbishop commented on qiskit/synthesis/two_qubit/two_qubit_decompose.py:
```diff
@@ -151,306 +151,56 @@ class TwoQubitWeylDecomposition:
-                logger.warning(
-                    "Requested fidelity different from actual by %s",
-                    self.calculated_fidelity - actual_fidelity,
-                )
```
Could consider moving this logging into the `__init__` wrapper? It would only have meaningful performance impact if debug logging is enabled.

mtreinish replied:
Yeah I can add it to `__init__` and also add the context manager back to the tests. 

My only concern around the logging is that for the transpiler when you enable debug logging on qiskit as a whole this can get super verbose. But, I think leaving it in the Python side is a good compromise because when we get the full default synthesis plugin migrated to rust we won't use the python side constructor for `TwoQubitWeylDecomposition` anymore in the transpiler so the debug messages will be only for debugging `TwoQubitWeylDecomposition` directly.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/11946/commits/03c4af53523f4fabbb72d82fe4a0bea61d09ecc4

levbishop commented on qiskit/synthesis/two_qubit/two_qubit_decompose.py:
```diff
@@ -467,6 +217,7 @@ def __repr__(self):
             + b64ascii
             + [
                 f"requested_fidelity={self.requested_fidelity},",
+                f"_specialization={self._inner_decomposition.specialization},"
```
the specializations wouldn't typically be in scope so the `__repr__` wouldn't normally be `eval()`-able without importing an internal API?

jakelishman replied:
I think that's quite normal - they're still importable, and the individual `repr` of any object doesn't usually include qualified access.  Most people don't `from numpy import array`, but `repr(numpy.array([1, 2, 3]))` is `array([1, 2, 3])`, for example.

levbishop replied:
```suggestion
                f"_specialization={self._inner_decomposition.specialization},",
```
(To make the formatting nicer)

levbishop replied:
I'm unsure on this. I agree that it shouldn't  be a requirement that everything can always eval without *any* imports, but seems unfriendly making folks figure out that they need to do something pretty non-obvious like `from qiskit._accelerate.two_qubit_decompose import Specialization` in order to use this functionality (which really only exists in the first place to help debugging when things are already confusing). Its unfortunate that the pyo3-generated `__repr__()` for the enum members doesn't include the full module path. Don't block merging based on doing something about this, but if you can think of a simple way to make it more user-friendly (without necessarily making the enum part of any public API) that would be good.

jakelishman replied:
> Its unfortunate that the pyo3-generated `__repr__()` for the enum members doesn't include the full module path.

What classes are you thinking of where this would be the case?  Python-object reprs aren't fully specified - the `x = eval(repr(x))` thing only ever _really_ applied to Python built-ins anyway, and not even those; there's nothing in `builtins` or `types` that you can import that can make `repr({}.keys())` eval'able, for example, despite it having the "eval-like" form.

Maybe I'm missing something about why the Python-space `__repr__` included a `from_bytes` in the first place, rather than being `f"{self.__class__.__name__}({self.unitary_matrix!r}, requested_fidelity={self.requested_fidelity!r})"` (plus the extra dummy kwargs if you wanted, though I'd argue they're perhaps more relevant for `__str__` than `__repr__`).  It seemed to me that the `from_bytes` method still recalculated all the same things as `__new__` anyway except for having the particular specialisation fixed, and since each class has its own `__new__` determined by `__init_subclass__`, we could have enforced the specialisation through that same mechanism, even in the factory-function `__new__` (i.e., at the point we check `if fidelity is None: instance = object.__new__(cls)`, we could include the same shortcircuit for `if cls is not TwoQubitWeylDecomposition`).

jakelishman replied:
Oh, and to answer the last bit about making it more user friendly: honestly, I'm not certain to what degree it's super necessary, but if we were very much concerned about it, we could maybe attach the `Specialization` enum type object as a class attribute on the Python-space type object (and potentially the Rust-space one), and then it'll appear in the documentation of the methods (which people _really_ should read before just copy-paste `eval`ing deep numerical code, I'd think).

levbishop replied:
Honestly I don't completely remember how I ended up with this version of repr. I do remember that it was a brute force rewrite after I originally attempted a couple versions similar to what you suggest and those weren't good enough for getting the consistent debugging I needed for intermittent test fails. Possibly it was the now-fixed numpy array2str bug that bit me. (I believe the subclasses did already enforce the specialization you're suggesting, by forcing fidelity to None in init_subclass.)

Anyway, attaching the enumeration asa class attribute seems reasonable to me. But if I'm the only one who cares about this then I'm ok to be outvoted and leave as is.

mtreinish replied:
I set the enum to a private class attr on the class and used that in the repr so it'll work with `eval()` even without an import: https://github.com/Qiskit/qiskit/pull/11946/commits/4dda4beb6682fac945dfee2d5a6f6a7621a5e076

levbishop reviewed (commented):
Looks good. Couple comments inline. Regarding pickling -- I think this change breaks unpickling from a previous qiskit ver, but my recollection is that users wanted the class pickleable for multiprocessing-type workloads (is that right?) rather than persistence (which AFAIK we only provide on best-effort basis) so that's probably fine.

jakelishman commented:
Pickle isn't really meant for any sort of persistence, and it definitely makes no guarantees about being loadable between different versions of a library - I wouldn't consider it a valid use to try and load a pickle in a changed Python environment, so I think we're fine on that count.

mtreinish committed:
Remove eigh() util function

mtreinish committed:
Revert unecessary changes to callers of TwoQubitWeylDecomposition

mtreinish committed:
Restore debug logging and add test assertions for it

mtreinish committed:
Merge branch 'rust-whales-transcription' of github.com:mtreinish/qiskit-core into rust-whales-transcription

jakelishman reviewed (dismissed):
Thanks for all of the changes, and for pulling us through this absolute quagmire of moving perhaps our most complex numerical code down into Rust space.  This looks good to me, and the minimal net changes to the test suite following the last set of commits is really encouraging.

I'm happy to approve, knowing that we're intending to do more tweaking and work on this in follow-ups to better take advantage of it being pure Rust now, but I'll leave it open just a shade to check if Lev wanted to comment again after his last round.

*edit*: Lev confirmed to me offline that he is looking again.

levbishop commented on crates/accelerate/src/two_qubit_decompose.rs:
```diff
@@ -180,20 +299,872 @@ fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) ->
+    // https://arxiv.org/pdf/2001.08343.pdf
+    SimaabEquiv,
+    SimabbEquiv,
+    SimabmbEquiv,
```
```suggestion
    fSimaabEquiv,
    fSimabbEquiv,
    fSimabmbEquiv,
```
The conventional name for the gate is "fSim".

mtreinish replied:
I did this because the lint rules didn't like the lower case "f". But I agree it's better to use the proper name; I'll just suppress that lint rule

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/11946/commits/ade9c8010bc358f6abe2bb8ee180790a24be2dd4

levbishop commented on qiskit/synthesis/two_qubit/two_qubit_decompose.py:
```diff
@@ -476,129 +240,27 @@ def __repr__(self):
-
-
-class TwoQubitWeylPartialSWAPEquiv(TwoQubitWeylDecomposition):
-    r""":math:`U \sim U_d(\alpha\pi/4, \alpha\pi/4, \alpha\pi/4) \sim \text{SWAP}^\alpha`
-    This gate binds 3 parameters, we make it canonical by setting:
-    :math:`K2_l = Id`.
```
Maybe its worth carrying this (and the other subclass docstrings), documenting the conventions we've chosen for selecting the canonical representative for each gate, into comments somewhere in the rust implementation, since they conventions are both somewhat arbitrary and fairly non-obvious?

mtreinish replied:
I can add it as a comment to specialization arm in the match statement in the rust code.

mtreinish replied:
I added the docstring as an inline comment in: https://github.com/Qiskit/qiskit/pull/11946/commits/76d2ccf42849e29d96d46e15dd1eb58c1802a7bf

levbishop commented on qiskit/synthesis/two_qubit/two_qubit_decompose.py:
```diff
@@ -476,129 +240,27 @@ def __repr__(self):
+        return cls(arr, fidelity=requested_fidelity, _specialization=_specialization)
 
     def __str__(self):
         pre = f"{self.__class__.__name__}(\n\t"
```
Now that this PR does away with the subclassing - this str() representation no longer directly tells which  specialization was applied (there are still clues from the form of the circuit chosen based on `euler_basis` and `weyl_gate`). Maybe would be nice to add an indication of the specialization back?

mtreinish replied:
Added it to `__str__` in: https://github.com/Qiskit/qiskit/pull/11946/commits/c52bffa7c071da01d1a9a21e0f15c73748617e84

levbishop reviewed (commented):
A few small comments to consider - I'm not dead-set on any of them.

mtreinish committed:
Add specialization to __str__

mtreinish committed:
Update qiskit/synthesis/two_qubit/two_qubit_decompose.py

Co-authored-by: Lev Bishop <18673315+levbishop@users.noreply.github.com>

mtreinish committed:
Add previous specialized class docstrings as inline rust code comments

mtreinish committed:
Rename fSim variants and suprress rustc warning about camel case

mtreinish committed:
Merge branch 'main' into rust-whales-transcription

mtreinish committed:
Update tests for correct specialization enum name

mtreinish committed:
Expose specialization enum via private class attr and use for __repr__

levbishop reviewed (approved):
This is great. 🚀

jakelishman reviewed (approved):
Thanks for all the work!
